include/swift/AST/Expr.h

```diff
@@ -476,7 +476,9 @@ class alignas(8) Expr {
   /// is not an archetype or dependent type.
   bool isStaticallyDerivedMetatype(
       llvm::function_ref<Type(const Expr *)> getType =
-          [](const Expr *E) -> Type { return E->getType(); }) const;
+          [](const Expr *E) -> Type { return E->getType(); },
+      llvm::function_ref<bool(const Expr *)> isTypeReference =
+          [](const Expr *E) { return E->isTypeReference(); }) const;
```

Comment:
Would it be cleaner if Expr::isTypeReference() took the getType callback, instead of passing two callbacks here?

Reply:
It already takes `getType` but it's insufficient because we need a way to retrieve declarations from constraint system for `UnresolvedDotExpr` and `UnresolvedMemberExpr`.

Reply:
Do you still need the Expr version then, and if so can you pass in an optional callback to look up overload choices too? It’s a complex piece of code and it’s suboptimal to repeat it twice I think

Reply:
Looks like `Expr::isStaticallyDerivedMetatype` is also used by SILGenApply and CodeCompletion besides CSApply.

Reply:
I'm going to commit the least ugly version of adding getDecl callback to `isTypeReference` I could come up with after I run tests locally.

Reply:
Awesome, thanks!

Reply:
Done. Added `getDecl` parameter to `isTypeRerefence` which defaults to `return nullptr` so I can use it in `ConstraintSystem::isTypeReference` to fetch declarations related to `UnresolvedDotExpr` and `UnresolvedMemberExpr`. WDYT?